Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function

pearu opened the pull request:
As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97825
* #98292
* #98288



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

pearu committed:
Support sparse return values in gradcheck input function

[ghstack-poisoned]

pearu committed:
Update on "Support sparse return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.

TODO: extend existing autograd tests by removing `to_dense(masked=)` method calls from the return values of gradcheck input functions.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented on torch/autograd/gradcheck.py:
```diff
@@ -616,12 +675,13 @@ def vjp_fn(grad_output):
                                    retain_graph=True, allow_unused=True)
     # Compute everything twice to check for nondeterminism (which we call reentrancy)
     if fast_mode:
-        vjps1 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output.clone(), v)
-        vjps2 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output.clone(), v)
+        vjps1 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output, v)
+        vjps2 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output, v)
```
There is no reason to clone when _get_analytical_vjps_wrt_specific_output only uses `output.shape`.

pearu commented on torch/autograd/gradcheck.py:
```diff
@@ -679,9 +739,10 @@ def get_analytical_jacobian(inputs, output, nondet_tol=0.0, grad_out=1.0):
     def vjp_fn(grad_output):
         return torch.autograd.grad(output, diff_input_list, grad_output,
                                    retain_graph=True, allow_unused=True)
+    sample_output = output if masked else _densify(output)
     # Compute everything twice to check for nondeterminism (which we call reentrancy)
-    vjps1 = _compute_analytical_jacobian_rows(vjp_fn, output.clone())
-    vjps2 = _compute_analytical_jacobian_rows(vjp_fn, output.clone())
+    vjps1 = _compute_analytical_jacobian_rows(vjp_fn, sample_output)
+    vjps2 = _compute_analytical_jacobian_rows(vjp_fn, sample_output)
```
No point in cloning `output` when `_compute_analytical_jacobian_rows` only calls `_zeros_like(sample_output)`.

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
@albanD @soulitzer this PR is ready for review. The two CI failures are unrelated to this PR.

soulitzer reviewed (commented):
Thanks for the addition, but I'm not sure we need this? Generally we've avoided adding more complexity in gradcheck by saying that alternative tensors (see nested tensors) would need to convert to dense tensors and back in order to use gradcheck.

soulitzer commented:
> enabling autograd-related opinfo tests for operations on sparse tensors

OpInfo has a `gradcheck_wrapper` field which can help with this.

> discovering bugs in the sparse support of PyTorch that otherwise exists but are unnoted as already evident from this PR (see test_sparse_mul, for instance).

Interesting. Maybe its worth investigating these failures first to see if they can be systematically caught another way and understand why converting to dense tensors can hinder the discovery of these bugs.

> Even though one may consider sparse tensors as alternative tensors (which is a rather unmotivating view, IMHO, but understandable)

I guess what I was trying to say is that its a trade off, gradcheck needs to be discriminatory wrt tensor types in order to minimize complexity. It's already handling a lot of things like complex, fast/slow, etc. The assumption here though is that the conversion to and back does not hide any bugs, and I believe that generally that is the case.

pearu commented:
> > discovering bugs in the sparse support of PyTorch that otherwise exists but are unnoted as already evident from this PR (see test_sparse_mul, for instance).
> 
> Interesting. Maybe its worth investigating these failures first to see if they can be systematically caught another way and understand why converting to dense tensors can hinder the discovery of these bugs.

Absolutely.

> > Even though one may consider sparse tensors as alternative tensors (which is a rather unmotivating view, IMHO, but understandable)
> 
> I guess what I was trying to say is that its a trade off, gradcheck needs to be discriminatory wrt tensor types in order to minimize complexity. It's already handling a lot of things like complex, fast/slow, etc. The assumption here though is that the conversion to and back does not hide any bugs, and I believe that generally that is the case.

IIUC, the main objective of gradcheck is to find bugs: if our assumptions on how the code works are correct, gradcheck would not be needed as it would always succeed.

Using Opinfo gradcheck_wrapper that applies conversions to/from dense tensors to satisfy the current gradcheck input conditions ("It's A Strided's Strided's Strided's World") may be practical in the short term but eventually, it is just about postponing the resolution of
https://github.com/pytorch/pytorch/blob/e61b8420010f43862a11ac250e8d9f24c8a75c4f/torch/autograd/gradcheck.py#L766-L769
Btw, there is no need to modify the analytical jacobian as the comment suggests. The default non-masked semantics of sparse tensors implies that sparse tensors and ops are semantically equivalent to strided tensors and torch ops. So, enabling sparse support in gradcheck would also enable discovering code pieces that violate the semantics equivalence while the conversion to strided tensors would hide such violations. That would be the main reason why we need this PR.

This PR makes two kinds of modifications to gradcheck.py:
1. workarounds to missing functionalities in sparse tensors support. For instance, sparse compressed tensors don't support elementary operations such as subtraction and division, and there exist bugs in scalar multiplication, also sparse tensors don't support `reshape`.
2. extending the gradcheck layout support by introducing helper functions such as `_get_values`, `_zeros_like`, `_get_layout_to_strided_permutation`.

When issues in 1. are resolved, the changeset in this PR will reduce.

The added complexity to gradcheck comes from the modifications in 2. but notice that the corresponding helper functions are no-ops or identity functions for strided tensors. The complexity of all other parts in gradcheck is basically unchanged (passing around `masked` kw arguments or improving exception messages or other trivial changes are not accounted for).

How about if I split this PR into smaller ghstack PRs that resolve issues in 1. so that the necessary changes to gradcheck will be minimized? Would that be an acceptable approach?

soulitzer commented:
Thanks for the clarification on the modifications here. I'm not sure if it is the right approach yet because I still think its worth thinking about some more things:

(1) what limitations are there with to_dense wrt testing coverage, and if there's other workarounds to achieve the same coverage, (2) is this still necessary to do if we could support sparse tensors at the OpInfo level instead, e.g. have an automatic way of wrapping input functions to do to_dense - investigating (1) would help determine if this is possible. (3) if we decide that yes, we do want to improve support for sparse in gradcheck (e.g. because using the to_dense wrapper limits testing coverage in a way that we cannot recover), how can we scale that to support other tensor types like nested tensor (could we factor some of this complexity out? - maybe some of that can be explained by your last comment, but I haven't read the PR closely yet)

> enabling sparse support in gradcheck would also enable discovering code pieces that violate the semantics equivalence while the conversion to strided tensors would hide such violations.

btw, this should just be a problem for slow gradcheck right? For slow gradcheck if I can see that if I want to pass in a basis vector as grad_output in a sparse way I cannot with the current approach. If I'm doing fast gradcheck I'm going to want to pass a random dense vector always anyway, so that does not seem problematic. Am I understanding the problem correctly?

If so, one dumb workaround is just to have a custom autograd Function as a wrapper around to_dense to recover the sparseness of zero-valued entries.

> but notice that the corresponding helper functions are no-ops or identity functions for strided tensors. 

True, but for someone who is not as familiar with the code, it may be hard to realize that immediately

pearu commented:
Re "(1) what limitations are there with to_dense wrt testing coverage":

One limitation comes from choosing the proper arguments to `to_dense`.
Namely, when wrapping a user-provided gradcheck input function with a function that applies `to_dense` to the input function result (say, via `gradcheck_wrapper`), there exist two ways to call `to_dense`: with `masked_grad=True` and `masked_grad=False` depending on the value of `masked` kw to `gradcheck`. If a wrong value for `masked_grad` is chosen, `gradcheck` will likely fail with a Jacobian mismatch error. With this PR, choosing `masked_grad` becomes unnecessary. Otherwise, we have to find a way to pass a proper value of `masked_grad` to `gradcheck_wrapper` depending on how `gradcheck` is called wrt `masked` argument.

Another limitation is that if `gradcheck` input function is allowed to return only strided tensors then `gradgradcheck` cannot support sparse tensors because `torch.autograd.grad` may return sparse tensors. I guess we could apply the proper `to_dense` call within the `gradgradcheck` generated input function to `gradcheck` but I have not tested it and I don't know its implications yet.

Re other questions, I don't have immediate answers because these require more investigation.

cpuhrsch reviewed (changes requested):
After further online discussion during our meeting we decided that the constraint here is that gradcheck must not be modified for this.

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Support sparse COO/CSR/CSC/BSR/BSC return values in gradcheck input function"

As in the title.

Previously, the input functions to gradcheck were expected to return strided tensors: say, when the result of an input function would be a sparse tensor, it had to be converted to strided via `.to_dense(masked=...)` method call. With this PR, the conversion of input function return values to strided tensors is not required anymore.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
@pytorchbot rebase
